### ISSUE TITLE
DHF 5.0.1 Smart Mastering RFE: logging of match candidate scoring

### ISSUE THREAD

**Issue by crism** [marklogic/marklogic-data-hub#2834]
#### The issue

Short description of the problem: I would like to be better able to debug my matching rules. The system is currently mostly opaque.

What behavior are you expecting? It would be nice to be able to hit some configuration switch, either in a flow description file or as a Gradle property, and get logging information on match candidate rules.

Ideally, I’d like to see something like:
```
id match contributed 20
name match contributed 10
family reduce role contributed -10
```
but I don’t believe that’s possible, the way queries are constructed for matching. In lieu of that, just seeing the queries that were run to find matches would be a good start, and what the total score was. Output could be in the log file or in auditing items in the database.

#### Tech details

Which Operating System are you using? CentOS 7

Which version of MarkLogic are you using? 9.0-9

Which version of the Data Hub Framework are you using? 5.0.1

Is this a QuickStart UI bug? If so, which browser are you using? N/A

**Comment by aajacobs** [comment]
+1

**Comment by rjrudin** [comment]
@crism I recommend expanding this to what trace events you'd want to see throughout the match and merge processes. Consider the simplest example - you have two documents that you're mastering, and you know they should be matched and merged. What trace events would you want to see throughout the process? For example (paraphrasing for each trace event here):

1. Constructing match query for URI: $uri
1. Constructed match query: $query
1. Found $n matches (and then list the URI and score of each match?)
1. Merging URI $uri with score $score
1. etc

I think trace events that describe the whole process would be very helpful for development (but far less so during big batch processes where the trace logging would be too noisy).

**Comment by crism** [comment]
When this option is enabled, I’d like to see, in the audit logs, for each match or notification (_i.e._, each triggered result):

 1. What the triggered match rule was.
 1. What the score was that caused the rule to trigger.
 1. What the query was that gave that score.
 1. Maybe which item was the focus and which ones were matches.

Your list is very similar, though procedural rather than result-oriented, but notably lacks which rule fired, which is really important for debugging.

**Comment by sbayatpur** [comment]
@crism Thanks for writing this out so clearly. We are planning to improve match candidate logging in the next feature release.

**Comment by sbayatpur** [comment]
Addressing this request with these 2 stories:
- https://project.marklogic.com/jira/browse/DHFPROD-2986
-https://project.marklogic.com/jira/browse/DHFPROD-2965

@crism let me know if you have feedback on these

**Comment by kkanthet** [comment]
Issue Fixed in 5.1.0 release
